                                        PYTHON PROGRAMMING

                                    MTN COTE D'IVOIRE DATA ANALYSIS

**RESEARCH OBJECTIVE**

Currently MTN Cote d'Ivoire would like to upgrade its technology infrastructure for its mobile users in Ivory Coast. Studying the given dataset, how does MTN Cote d'Ivoire go about the upgrade of its infrastructure strategy within the given cities?

In [44]:
#we begin by first importing pandas and numpy
import pandas as pd
import numpy as np


IMPORTING THE DATA TABLES

In [45]:
#loadinng our data tables
# CRD Decription table
cdr = pd.read_excel('CDR_description.xlsx')
cdr = pd.DataFrame(cdr)
cdr

Column name                                        Description   Format
0          PRODUCT                                       Voice or SMS   String
1            VALUE                                      Billing price  Integer
2        DATE_TIME               Time in format yyyy-MM-dd hh:mm:ss.0   String
3     CELL_ON_SITE  Which cell in the site was used (not needed here)  Integer
4  DW_A_NUMBER_INT  Anonymized phone number of the person for whic...   String
5  DW_B_NUMBER_INT        Anonymized phone number of the counterparty   String
6        COUNTRY_A                  Country of party A (useless here)   String
7        COUNTRY_B                  Country of party B (useless here)   String
8          CELL_ID                                     ID of the cell   String
9          SITE_ID                                     ID of the SITE   String

In [46]:
#cells geo description table
cellgeodesc = pd.read_excel('cells_geo_description.xlsx')
cellgeodesc = pd.DataFrame(cellgeodesc)
cellgeodesc

Column name                                        Description  Format
0         VILLES                                               City  String
1         STATUS                                  In Service or not  String
2   LOCALISATION                                  in ABIDJAN or not  String
3     DECOUPZONE                                  Geographical Zone  String
4       ZONENAME                                       Name of Zone  String
5      LONGITUDE                                          Longitude   Float
6       LATITUDE                                           Latitude   Float
7         REGION                                             Region  String
8           AREA                                               Area  String
9        CELL_ID                                     ID of the cell  String
10     SITE_CODE  Site (there are several cells per site, severa...  String

In [47]:
#cells geo table
cellgeo = pd.read_csv('cells_geo.csv', delimiter = r';', engine ='python')
cellgeo = pd.DataFrame(cellgeo)
cellgeo

Unnamed: 0    VILLES      STATUS LOCALISATION    DECOUPZONE  \
0              0    ADJAME  In Service      ABIDJAN  "Abidjan_EST   
1              1    ADJAME  In Service      ABIDJAN  "Abidjan_EST   
2              2    ADJAME  In Service      ABIDJAN  "Abidjan_EST   
3              3   ASSINIE  In Service    INTERIEUR     Grand-EST   
4              4   ASSINIE  In Service    INTERIEUR     Grand-EST   
...          ...       ...         ...          ...           ...   
3969        3969  ZUENOULA  In Service    INTERIEUR    Grand-NORD   
3970        3970  ZUENOULA  In Service    INTERIEUR    Grand-NORD   
3971        3971  ZUENOULA  In Service    INTERIEUR    Grand-NORD   
3972        3972  ZUENOULA  In Service    INTERIEUR    Grand-NORD   
3973        3973  ZUENOULA  In Service    INTERIEUR    Grand-NORD   

                 ZONENAME  LONGITUDE  LATITUDE     REGION          AREA  \
0                AGBANNEW  -4.000667  5.368194     LAGUNE           NaN   
1                AGBANNEW  -4.000667  5.368194     LAGUNE           NaN   
2                AGBANNEW  -4.000667  5.368194     LAGUNE           NaN   
3                ASSINIE"  -3.488467  5.158761  SUD COMOE           NaN   
4                ASSINIE"  -3.488467  5.158761  SUD COMOE           NaN   
...                   ...        ...       ...        ...           ...   
3969  ZONE A - Grand-NORD  -6.039694  7.428611   MARAHOUE  CENTRE NORTH   
3970  ZONE A - Grand-NORD  -6.039694  7.428611   MARAHOUE  CENTRE NORTH   
3971  ZONE A - Grand-NORD  -6.065450  7.444592   MARAHOUE  CENTRE NORTH   
3972  ZONE A - Grand-NORD  -6.065450  7.444592   MARAHOUE  CENTRE NORTH   
3973  ZONE A - Grand-NORD  -6.065450  7.444592   MARAHOUE  CENTRE NORTH   

         CELL_ID   SITE_CODE  
0     fbb2e75af3  0bb492ef17  
1     1502501d78  0bb492ef17  
2     0c2e182a0c  0bb492ef17  
3     69bdfc69fd  8a1e6ce31f  
4     7a1029d2ab  8a1e6ce31f  
...          ...         ...  
3969  1cfc2d7ab1  f91ed8d2e9  
3970  3ae1708153  f91ed8d2e9  
3971  cb71f8a29c  61bee9c70c  
3972  d4a6a0b645  61bee9c70c  
3973  f9f3e42589  61bee9c70c  

[3974 rows x 12 columns]

In [48]:
Telcom1 = pd.read_csv('Telcom_dataset.csv')
Telcom1 = pd.DataFrame(Telcom1)
Telcom1

PRODUTC  VALUE               DATETIME CELL_ON_SITE  DW_A_NUMBER_INT  \
0      Voice      0  2012-05-06 23:04:37.0            1  49fae048b547780   
1        sms      0  2012-05-06 23:05:25.0            0  7a54a6eaee19e65   
2        sms     25  2012-05-06 23:05:41.0            0  6008887b0207d82   
3      Voice     56  2012-05-06 23:05:42.0            1  7370032e20f4e96   
4      Voice     14  2012-05-06 23:06:29.0            1  966ba4656e38e62   
...      ...    ...                    ...          ...              ...   
4996     sms    100  2012-05-07 00:01:54.0            0  92e367d671f45c5   
4997   Voice      1  2012-05-07 00:01:54.0            1  83b44b18238e602   
4998   Voice      0  2012-05-07 00:01:54.0            1  49fae048b547780   
4999   Voice      0  2012-05-07 00:01:54.0            0  a1613f1a992716a   
5000   Voice      7  2012-05-07 00:01:54.0            2  1670a5f7c22f4ed   

      DW_B_NUMBER_INT COUNTRY_A COUNTRY_B     CELL_ID     SITE_ID  
0     1a49f8b0ded9de6     19e2e     19e2e  /TJNe+Mmtu  /TJNe+Mmtu  
1     93343ba6e41af95     19e2e     19e2e  /+cKZKIp41  /+cKZKIp41  
2     7aacc9709ce0a20     19e2e     19e2e  /u0FSD+ahi  /u0FSD+ahi  
3     6af4a60e74c91d3     19e2e     19e2e  +cNeJzsTp3  +cNeJzsTp3  
4     02aa736ee555d5c     19e2e     19e2e  051KhYDCpv  051KhYDCpv  
...               ...       ...       ...         ...         ...  
4996  38811081cf82c90     19e2e     OTHER  0c2371d9bc  1e751d08a4  
4997  6ec63d4aa5de5dd     19e2e     19e2e  5fb791cf56  65efe2bc62  
4998  1a49f8b0ded9de6     19e2e     19e2e  7773627c12  9ef9616350  
4999  bd663f07b6c8bcd     19e2e     19e2e  3f4af49d6c  969b8d0e39  
5000  3600efd094f2225     19e2e     19e2e  6f4d9b77a5  ebfc58af70  

[5001 rows x 10 columns]

In [49]:
#Updating the column names to make the uniform aand correct tyoing errors
Telcom1.columns =['PRODUCT','VALUE','DATE_TIME','CELL_ON_SITE','DW_A_NUMBER','DW_B_NUMBER','COUNTRY_A','COUNTRY_B','CELL_ID','SITE_ID']
Telcom1

PRODUCT  VALUE              DATE_TIME CELL_ON_SITE      DW_A_NUMBER  \
0      Voice      0  2012-05-06 23:04:37.0            1  49fae048b547780   
1        sms      0  2012-05-06 23:05:25.0            0  7a54a6eaee19e65   
2        sms     25  2012-05-06 23:05:41.0            0  6008887b0207d82   
3      Voice     56  2012-05-06 23:05:42.0            1  7370032e20f4e96   
4      Voice     14  2012-05-06 23:06:29.0            1  966ba4656e38e62   
...      ...    ...                    ...          ...              ...   
4996     sms    100  2012-05-07 00:01:54.0            0  92e367d671f45c5   
4997   Voice      1  2012-05-07 00:01:54.0            1  83b44b18238e602   
4998   Voice      0  2012-05-07 00:01:54.0            1  49fae048b547780   
4999   Voice      0  2012-05-07 00:01:54.0            0  a1613f1a992716a   
5000   Voice      7  2012-05-07 00:01:54.0            2  1670a5f7c22f4ed   

          DW_B_NUMBER COUNTRY_A COUNTRY_B     CELL_ID     SITE_ID  
0     1a49f8b0ded9de6     19e2e     19e2e  /TJNe+Mmtu  /TJNe+Mmtu  
1     93343ba6e41af95     19e2e     19e2e  /+cKZKIp41  /+cKZKIp41  
2     7aacc9709ce0a20     19e2e     19e2e  /u0FSD+ahi  /u0FSD+ahi  
3     6af4a60e74c91d3     19e2e     19e2e  +cNeJzsTp3  +cNeJzsTp3  
4     02aa736ee555d5c     19e2e     19e2e  051KhYDCpv  051KhYDCpv  
...               ...       ...       ...         ...         ...  
4996  38811081cf82c90     19e2e     OTHER  0c2371d9bc  1e751d08a4  
4997  6ec63d4aa5de5dd     19e2e     19e2e  5fb791cf56  65efe2bc62  
4998  1a49f8b0ded9de6     19e2e     19e2e  7773627c12  9ef9616350  
4999  bd663f07b6c8bcd     19e2e     19e2e  3f4af49d6c  969b8d0e39  
5000  3600efd094f2225     19e2e     19e2e  6f4d9b77a5  ebfc58af70  

[5001 rows x 10 columns]

In [50]:
Telcom2 = pd.read_csv('Telcom_dataset2.csv')
Telcom2 = pd.DataFrame(Telcom2)
Telcom2

PRODUCT  VALUE              DATE_TIME CELL_ON_SITE      DW_A_NUMBER  \
0        sms      0  2012-05-07 23:02:06.0            2  32255675c3b6fe3   
1      Voice      0  2012-05-07 23:03:44.0            1  ff16d6ac74c1c64   
2        sms      0  2012-05-07 23:04:06.0            0  98d9f2fb5ff624b   
3      Voice     37  2012-05-07 23:04:48.0            1  3f9c8a33ec7be1f   
4        sms     25  2012-05-07 23:04:58.0            1  7bcccc03adf3864   
...      ...    ...                    ...          ...              ...   
4996     sms      0  2012-05-08 00:01:53.0            8  22ffbbb8ac8cde1   
4997     sms     25  2012-05-08 00:01:53.0            0  bcf512a3542d42c   
4998     sms      0  2012-05-08 00:01:53.0            0  05e3c3d41628a9f   
4999    data      0  2012-05-08 00:01:53.0            _  86a4dbf018f6f24   
5000   Voice      0  2012-05-08 00:01:53.0            0  a03c9837c62a5b9   

          DW_B_NUMBER COUNTRY_A COUNTRY_B     CELL_ID     SITE_ID  
0     1aa7ae01386e9fb     19e2e     19e2e  +854AcBQT2  +854AcBQT2  
1     87d87309c15a8c9     19e2e     19e2e  +854AcBQT2  +854AcBQT2  
2     93343ba6e41af95     19e2e     19e2e  +laSrk7g6q  +laSrk7g6q  
3     7d05a38f89c49d8     19e2e     19e2e  /xvRrCVKoQ  /xvRrCVKoQ  
4     56a5d5430d2fa86     19e2e     19e2e  +bO+qkSonO  +bO+qkSonO  
...               ...       ...       ...         ...         ...  
4996  8db839a0c08bd54     19e2e     19e2e  896459aace  b7d00e59e4  
4997  5a2848d6f983c31     19e2e     19e2e  7e90b35166  211015b14d  
4998  7185a95077eae9e     19e2e     19e2e  fe70e431df  c7ea989264  
4999  5a4dc9e82deb309     19e2e     OTHER  ffa6759bb2         NaN  
5000  de590e98879fa37     19e2e     19e2e  8630058aca  b47a998a3e  

[5001 rows x 10 columns]

In [51]:
Telcom3 =  pd.read_csv('Telcom_dataset.csv')
Telcom3 = pd.DataFrame(Telcom3)
Telcom3

PRODUTC  VALUE               DATETIME CELL_ON_SITE  DW_A_NUMBER_INT  \
0      Voice      0  2012-05-06 23:04:37.0            1  49fae048b547780   
1        sms      0  2012-05-06 23:05:25.0            0  7a54a6eaee19e65   
2        sms     25  2012-05-06 23:05:41.0            0  6008887b0207d82   
3      Voice     56  2012-05-06 23:05:42.0            1  7370032e20f4e96   
4      Voice     14  2012-05-06 23:06:29.0            1  966ba4656e38e62   
...      ...    ...                    ...          ...              ...   
4996     sms    100  2012-05-07 00:01:54.0            0  92e367d671f45c5   
4997   Voice      1  2012-05-07 00:01:54.0            1  83b44b18238e602   
4998   Voice      0  2012-05-07 00:01:54.0            1  49fae048b547780   
4999   Voice      0  2012-05-07 00:01:54.0            0  a1613f1a992716a   
5000   Voice      7  2012-05-07 00:01:54.0            2  1670a5f7c22f4ed   

      DW_B_NUMBER_INT COUNTRY_A COUNTRY_B     CELL_ID     SITE_ID  
0     1a49f8b0ded9de6     19e2e     19e2e  /TJNe+Mmtu  /TJNe+Mmtu  
1     93343ba6e41af95     19e2e     19e2e  /+cKZKIp41  /+cKZKIp41  
2     7aacc9709ce0a20     19e2e     19e2e  /u0FSD+ahi  /u0FSD+ahi  
3     6af4a60e74c91d3     19e2e     19e2e  +cNeJzsTp3  +cNeJzsTp3  
4     02aa736ee555d5c     19e2e     19e2e  051KhYDCpv  051KhYDCpv  
...               ...       ...       ...         ...         ...  
4996  38811081cf82c90     19e2e     OTHER  0c2371d9bc  1e751d08a4  
4997  6ec63d4aa5de5dd     19e2e     19e2e  5fb791cf56  65efe2bc62  
4998  1a49f8b0ded9de6     19e2e     19e2e  7773627c12  9ef9616350  
4999  bd663f07b6c8bcd     19e2e     19e2e  3f4af49d6c  969b8d0e39  
5000  3600efd094f2225     19e2e     19e2e  6f4d9b77a5  ebfc58af70  

[5001 rows x 10 columns]

In [52]:
#Updating the column names to make the uniform aand correct tyoing errors
Telcom3.columns =['PRODUCT','VALUE','DATE_TIME','CELL_ON_SITE','DW_A_NUMBER','DW_B_NUMBER','COUNTRY_A','COUNTRY_B','CELL_ID','SITE_ID']
Telcom3

PRODUCT  VALUE              DATE_TIME CELL_ON_SITE      DW_A_NUMBER  \
0      Voice      0  2012-05-06 23:04:37.0            1  49fae048b547780   
1        sms      0  2012-05-06 23:05:25.0            0  7a54a6eaee19e65   
2        sms     25  2012-05-06 23:05:41.0            0  6008887b0207d82   
3      Voice     56  2012-05-06 23:05:42.0            1  7370032e20f4e96   
4      Voice     14  2012-05-06 23:06:29.0            1  966ba4656e38e62   
...      ...    ...                    ...          ...              ...   
4996     sms    100  2012-05-07 00:01:54.0            0  92e367d671f45c5   
4997   Voice      1  2012-05-07 00:01:54.0            1  83b44b18238e602   
4998   Voice      0  2012-05-07 00:01:54.0            1  49fae048b547780   
4999   Voice      0  2012-05-07 00:01:54.0            0  a1613f1a992716a   
5000   Voice      7  2012-05-07 00:01:54.0            2  1670a5f7c22f4ed   

          DW_B_NUMBER COUNTRY_A COUNTRY_B     CELL_ID     SITE_ID  
0     1a49f8b0ded9de6     19e2e     19e2e  /TJNe+Mmtu  /TJNe+Mmtu  
1     93343ba6e41af95     19e2e     19e2e  /+cKZKIp41  /+cKZKIp41  
2     7aacc9709ce0a20     19e2e     19e2e  /u0FSD+ahi  /u0FSD+ahi  
3     6af4a60e74c91d3     19e2e     19e2e  +cNeJzsTp3  +cNeJzsTp3  
4     02aa736ee555d5c     19e2e     19e2e  051KhYDCpv  051KhYDCpv  
...               ...       ...       ...         ...         ...  
4996  38811081cf82c90     19e2e     OTHER  0c2371d9bc  1e751d08a4  
4997  6ec63d4aa5de5dd     19e2e     19e2e  5fb791cf56  65efe2bc62  
4998  1a49f8b0ded9de6     19e2e     19e2e  7773627c12  9ef9616350  
4999  bd663f07b6c8bcd     19e2e     19e2e  3f4af49d6c  969b8d0e39  
5000  3600efd094f2225     19e2e     19e2e  6f4d9b77a5  ebfc58af70  

[5001 rows x 10 columns]

Concatinating the 3 telcom datast tables as the have the same information

In [53]:
new = pd.concat([Telcom1, Telcom2])
new

PRODUCT  VALUE              DATE_TIME CELL_ON_SITE      DW_A_NUMBER  \
0      Voice      0  2012-05-06 23:04:37.0            1  49fae048b547780   
1        sms      0  2012-05-06 23:05:25.0            0  7a54a6eaee19e65   
2        sms     25  2012-05-06 23:05:41.0            0  6008887b0207d82   
3      Voice     56  2012-05-06 23:05:42.0            1  7370032e20f4e96   
4      Voice     14  2012-05-06 23:06:29.0            1  966ba4656e38e62   
...      ...    ...                    ...          ...              ...   
4996     sms      0  2012-05-08 00:01:53.0            8  22ffbbb8ac8cde1   
4997     sms     25  2012-05-08 00:01:53.0            0  bcf512a3542d42c   
4998     sms      0  2012-05-08 00:01:53.0            0  05e3c3d41628a9f   
4999    data      0  2012-05-08 00:01:53.0            _  86a4dbf018f6f24   
5000   Voice      0  2012-05-08 00:01:53.0            0  a03c9837c62a5b9   

          DW_B_NUMBER COUNTRY_A COUNTRY_B     CELL_ID     SITE_ID  
0     1a49f8b0ded9de6     19e2e     19e2e  /TJNe+Mmtu  /TJNe+Mmtu  
1     93343ba6e41af95     19e2e     19e2e  /+cKZKIp41  /+cKZKIp41  
2     7aacc9709ce0a20     19e2e     19e2e  /u0FSD+ahi  /u0FSD+ahi  
3     6af4a60e74c91d3     19e2e     19e2e  +cNeJzsTp3  +cNeJzsTp3  
4     02aa736ee555d5c     19e2e     19e2e  051KhYDCpv  051KhYDCpv  
...               ...       ...       ...         ...         ...  
4996  8db839a0c08bd54     19e2e     19e2e  896459aace  b7d00e59e4  
4997  5a2848d6f983c31     19e2e     19e2e  7e90b35166  211015b14d  
4998  7185a95077eae9e     19e2e     19e2e  fe70e431df  c7ea989264  
4999  5a4dc9e82deb309     19e2e     OTHER  ffa6759bb2         NaN  
5000  de590e98879fa37     19e2e     19e2e  8630058aca  b47a998a3e  

[10002 rows x 10 columns]

In [54]:
dftel = pd.concat([new, Telcom3])
dftel

PRODUCT  VALUE              DATE_TIME CELL_ON_SITE      DW_A_NUMBER  \
0      Voice      0  2012-05-06 23:04:37.0            1  49fae048b547780   
1        sms      0  2012-05-06 23:05:25.0            0  7a54a6eaee19e65   
2        sms     25  2012-05-06 23:05:41.0            0  6008887b0207d82   
3      Voice     56  2012-05-06 23:05:42.0            1  7370032e20f4e96   
4      Voice     14  2012-05-06 23:06:29.0            1  966ba4656e38e62   
...      ...    ...                    ...          ...              ...   
4996     sms    100  2012-05-07 00:01:54.0            0  92e367d671f45c5   
4997   Voice      1  2012-05-07 00:01:54.0            1  83b44b18238e602   
4998   Voice      0  2012-05-07 00:01:54.0            1  49fae048b547780   
4999   Voice      0  2012-05-07 00:01:54.0            0  a1613f1a992716a   
5000   Voice      7  2012-05-07 00:01:54.0            2  1670a5f7c22f4ed   

          DW_B_NUMBER COUNTRY_A COUNTRY_B     CELL_ID     SITE_ID  
0     1a49f8b0ded9de6     19e2e     19e2e  /TJNe+Mmtu  /TJNe+Mmtu  
1     93343ba6e41af95     19e2e     19e2e  /+cKZKIp41  /+cKZKIp41  
2     7aacc9709ce0a20     19e2e     19e2e  /u0FSD+ahi  /u0FSD+ahi  
3     6af4a60e74c91d3     19e2e     19e2e  +cNeJzsTp3  +cNeJzsTp3  
4     02aa736ee555d5c     19e2e     19e2e  051KhYDCpv  051KhYDCpv  
...               ...       ...       ...         ...         ...  
4996  38811081cf82c90     19e2e     OTHER  0c2371d9bc  1e751d08a4  
4997  6ec63d4aa5de5dd     19e2e     19e2e  5fb791cf56  65efe2bc62  
4998  1a49f8b0ded9de6     19e2e     19e2e  7773627c12  9ef9616350  
4999  bd663f07b6c8bcd     19e2e     19e2e  3f4af49d6c  969b8d0e39  
5000  3600efd094f2225     19e2e     19e2e  6f4d9b77a5  ebfc58af70  

[15003 rows x 10 columns]

DATA UNDERSTANDING

In [55]:
#information on the cell geographics table
cellgeo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3974 entries, 0 to 3973
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    3974 non-null   int64  
 1   VILLES        3974 non-null   object 
 2   STATUS        3907 non-null   object 
 3   LOCALISATION  3974 non-null   object 
 4   DECOUPZONE    3974 non-null   object 
 5   ZONENAME      3968 non-null   object 
 6   LONGITUDE     3974 non-null   float64
 7   LATITUDE      3974 non-null   float64
 8   REGION        3974 non-null   object 
 9   AREA          3951 non-null   object 
 10  CELL_ID       3974 non-null   object 
 11  SITE_CODE     3974 non-null   object 
dtypes: float64(2), int64(1), object(9)
memory usage: 372.7+ KB


the table above has 3,974 rows in 12 columns.
Status column has 67 missing entries
zone name has 6 missing value
area column has 23 missing entries


In [56]:
#Information on the telcom datasets
dftel.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15003 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   PRODUCT       15003 non-null  object
 1   VALUE         15003 non-null  int64 
 2   DATE_TIME     15003 non-null  object
 3   CELL_ON_SITE  15003 non-null  object
 4   DW_A_NUMBER   15003 non-null  object
 5   DW_B_NUMBER   15003 non-null  object
 6   COUNTRY_A     15003 non-null  object
 7   COUNTRY_B     15003 non-null  object
 8   CELL_ID       15003 non-null  object
 9   SITE_ID       12895 non-null  object
dtypes: int64(1), object(9)
memory usage: 1.3+ MB


the telcom dataset has 15,003 rows and 9 columns
the site_id column has 2,108 missing vavues


In [57]:
#drop non required colomns
dftel = dftel.drop(columns=['CELL_ON_SITE'])
dftel = dftel.drop(columns=['COUNTRY_A'])
dftel = dftel.drop(columns=['COUNTRY_B'])
cellgeo = cellgeo.drop(columns = ['Unnamed: 0'])

DATA CLEANING

dropping entriels with null values

In [58]:
dftel.dropna(inplace = True)

In [59]:
cellgeo.dropna(inplace = True)


In [60]:
#set the date/time format to ensure all date/time entries are in the right format
dftel['DATE_TIME'] = pd.to_datetime(dftel['DATE_TIME'])
dftel.head(5)


PRODUCT  VALUE           DATE_TIME      DW_A_NUMBER      DW_B_NUMBER  \
0   Voice      0 2012-05-06 23:04:37  49fae048b547780  1a49f8b0ded9de6   
1     sms      0 2012-05-06 23:05:25  7a54a6eaee19e65  93343ba6e41af95   
2     sms     25 2012-05-06 23:05:41  6008887b0207d82  7aacc9709ce0a20   
3   Voice     56 2012-05-06 23:05:42  7370032e20f4e96  6af4a60e74c91d3   
4   Voice     14 2012-05-06 23:06:29  966ba4656e38e62  02aa736ee555d5c   

      CELL_ID     SITE_ID  
0  /TJNe+Mmtu  /TJNe+Mmtu  
1  /+cKZKIp41  /+cKZKIp41  
2  /u0FSD+ahi  /u0FSD+ahi  
3  +cNeJzsTp3  +cNeJzsTp3  
4  051KhYDCpv  051KhYDCpv

In [61]:
#check for duplicates
print(cellgeo.duplicated())

23      False
24      False
25      False
26      False
27      False
        ...  
3969    False
3970    False
3971    False
3972    False
3973    False
Length: 3880, dtype: bool


In [62]:
#Check for duplicates
dftel.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
4996     True
4997     True
4998     True
4999     True
5000     True
Length: 12895, dtype: bool

DATA ANALYSIS

Grouping

In [63]:
#Groupint the cell geography table by city
villes = cellgeo.groupby('VILLES')
villes.first()

STATUS LOCALISATION    DECOUPZONE               ZONENAME  \
VILLES                                                                        
ABENGOUROU     In Service    INTERIEUR     Grand-EST     ZONE C - Grand-EST   
ABOBO          In Service      ABIDJAN  Abidjan_NORD  ZONE E - Abidjan_NORD   
ABOISSO        In Service    INTERIEUR     Grand-EST     ZONE C - Grand-EST   
ABOISSO COMOE  In Service    INTERIEUR     Grand-EST     ZONE C - Grand-EST   
ABONGOUA       In Service    INTERIEUR     Grand-EST     ZONE C - Grand-EST   
...                   ...          ...           ...                    ...   
ZATTA          In Service    INTERIEUR    Grand-NORD    ZONE B - Grand-NORD   
ZIKISSO        In Service    INTERIEUR     Grand-SUD     ZONE B - Grand-SUD   
ZON-HOUNIEN    In Service    INTERIEUR   Grand-OUEST   ZONE D - Grand-OUEST   
ZOUKOUGBEU     In Service    INTERIEUR   Grand-OUEST   ZONE D - Grand-OUEST   
ZUENOULA       In Service    INTERIEUR    Grand-NORD    ZONE A - Grand-NORD   

               LONGITUDE  LATITUDE          REGION           AREA     CELL_ID  \
VILLES                                                                          
ABENGOUROU     -3.514833  6.722333     MOYEN COMOE           EAST  8cdfc1bc65   
ABOBO          -4.014139  5.433500          LAGUNE  ABIDJAN NORTH  f740406877   
ABOISSO        -0.000000  0.000000       SUD COMOE  ABIDJAN SOUTH  b4d8c16d9e   
ABOISSO COMOE  -3.396861  5.734500          LAGUNE  ABIDJAN NORTH  ff4372eafb   
ABONGOUA       -3.548722  6.247083     MOYEN COMOE           EAST  d1a1270467   
...                  ...       ...             ...            ...         ...   
ZATTA          -5.415614  6.905422            LACS   CENTRE NORTH  1d41c565fc   
ZIKISSO        -5.698833  6.057278     SUD BANDAMA     SOUTH WEST  733df73327   
ZON-HOUNIEN    -8.205167  6.920500    18 MONTAGNES     NORTH WEST  a3e5534191   
ZOUKOUGBEU     -6.865694  6.761306  HAUT SASSANDRA     NORTH WEST  317ba2e3e8   
ZUENOULA       -6.138956  7.617553        MARAHOUE   CENTRE NORTH  8cde542067   

                SITE_CODE  
VILLES                     
ABENGOUROU     6e87d6cb59  
ABOBO          0f34c36ac5  
ABOISSO        38c5c1de07  
ABOISSO COMOE  640427c4c9  
ABONGOUA       a8d8462dda  
...                   ...  
ZATTA          c644d1ce4c  
ZIKISSO        211015b14d  
ZON-HOUNIEN    32b8357a82  
ZOUKOUGBEU     443ac3fc19  
ZUENOULA       43adae9a57  

[501 rows x 10 columns]

In [64]:
#use count() function to determine the number of cell activities recorded per city
villescount = villes.count()

In [65]:
# sort the grouping count in descending to see cities that had the most cell ussage in the 3 days.
dv = villescount.sort_values('CELL_ID', ascending = False)
#use head() to get the top 10 cities
dv.head(10)

STATUS  LOCALISATION  DECOUPZONE  ZONENAME  LONGITUDE  LATITUDE  \
VILLES                                                                          
COCODY           354           354         354       354        354       354   
YOPOUGON         289           289         289       289        289       289   
ABOBO            244           244         244       244        244       244   
MARCORY          141           141         141       141        141       141   
TREICHVILLE      132           132         132       132        132       132   
YAMOUSSOUKRO     104           104         104       104        104       104   
ADJAME           104           104         104       104        104       104   
PLATEAU          100           100         100       100        100       100   
KOUMASSI          91            91          91        91         91        91   
BOUAKE            65            65          65        65         65        65   

              REGION  AREA  CELL_ID  SITE_CODE  
VILLES                                          
COCODY           354   354      354        354  
YOPOUGON         289   289      289        289  
ABOBO            244   244      244        244  
MARCORY          141   141      141        141  
TREICHVILLE      132   132      132        132  
YAMOUSSOUKRO     104   104      104        104  
ADJAME           104   104      104        104  
PLATEAU          100   100      100        100  
KOUMASSI          91    91       91         91  
BOUAKE            65    65       65         65

In [66]:
# describing the group count to see the statistics of all data collected in the three days.
dv.describe()


STATUS  LOCALISATION  DECOUPZONE    ZONENAME   LONGITUDE  \
count  501.000000    501.000000  501.000000  501.000000  501.000000   
mean     7.744511      7.744511    7.744511    7.744511    7.744511   
std     26.510047     26.510047   26.510047   26.510047   26.510047   
min      1.000000      1.000000    1.000000    1.000000    1.000000   
25%      3.000000      3.000000    3.000000    3.000000    3.000000   
50%      3.000000      3.000000    3.000000    3.000000    3.000000   
75%      3.000000      3.000000    3.000000    3.000000    3.000000   
max    354.000000    354.000000  354.000000  354.000000  354.000000   

         LATITUDE      REGION        AREA     CELL_ID   SITE_CODE  
count  501.000000  501.000000  501.000000  501.000000  501.000000  
mean     7.744511    7.744511    7.744511    7.744511    7.744511  
std     26.510047   26.510047   26.510047   26.510047   26.510047  
min      1.000000    1.000000    1.000000    1.000000    1.000000  
25%      3.000000    3.000000    3.000000    3.000000    3.000000  
50%      3.000000    3.000000    3.000000    3.000000    3.000000  
75%      3.000000    3.000000    3.000000    3.000000    3.000000  
max    354.000000  354.000000  354.000000  354.000000  354.000000

In [67]:
#use tail() to filter the cities that recorded the least number of cell phone activity.
dv.tail(5)

STATUS  LOCALISATION  DECOUPZONE  ZONENAME  LONGITUDE  LATITUDE  \
VILLES                                                                      
KOTOULA        2             2           2         2          2         2   
KRIKOREA       2             2           2         2          2         2   
MEMNI          2             2           2         2          2         2   
OUANGOLO       1             1           1         1          1         1   
DANANON        1             1           1         1          1         1   

          REGION  AREA  CELL_ID  SITE_CODE  
VILLES                                      
KOTOULA        2     2        2          2  
KRIKOREA       2     2        2          2  
MEMNI          2     2        2          2  
OUANGOLO       1     1        1          1  
DANANON        1     1        1          1

In [68]:
#group the telcom data by products to see the most used product between voice,sms and data
tel = dftel.groupby('PRODUCT')
tel.first()

VALUE           DATE_TIME      DW_A_NUMBER      DW_B_NUMBER  \
PRODUCT                                                                
Voice        0 2012-05-06 23:04:37  49fae048b547780  1a49f8b0ded9de6   
data         0 2012-05-06 23:07:26  f71825ca6e623f9  5a4dc9e82deb309   
sms          0 2012-05-06 23:05:25  7a54a6eaee19e65  93343ba6e41af95   

            CELL_ID     SITE_ID  
PRODUCT                          
Voice    /TJNe+Mmtu  /TJNe+Mmtu  
data     /lrUXVGVcN  /lrUXVGVcN  
sms      /+cKZKIp41  /+cKZKIp41

In [69]:
#use count() to see the number of times each product was usedin the three days
telcount = tel.count()
telcount

VALUE  DATE_TIME  DW_A_NUMBER  DW_B_NUMBER  CELL_ID  SITE_ID
PRODUCT                                                              
Voice     6342       6342         6342         6342     6342     6342
data       274        274          274          274      274      274
sms       6279       6279         6279         6279     6279     6279

In [70]:
#obtaining the total bill value from each product offered by MTN
tel2 = dftel.groupby('PRODUCT').sum('VALUE')
print(tel2)

          VALUE
PRODUCT        
Voice    321224
data       1999
sms       39659


In [71]:
#grouping by those in ABIDJAN and those that are not in Abidjan
local = cellgeo.groupby('LOCALISATION')
local.count()

VILLES  STATUS  DECOUPZONE  ZONENAME  LONGITUDE  LATITUDE  \
LOCALISATION                                                              
ABIDJAN         1619    1619        1619      1619       1619      1619   
INTERIEUR       2261    2261        2261      2261       2261      2261   

              REGION  AREA  CELL_ID  SITE_CODE  
LOCALISATION                                    
ABIDJAN         1619  1619     1619       1619  
INTERIEUR       2261  2261     2261       2261

In [72]:
#grouping by siteid
site = dftel.groupby('SITE_ID')

In [73]:
#use count() to see how many times each site id was used in the three days.
# ordered in descending order
sitecount = site.count()
siteid = sitecount.sort_values('VALUE',ascending = False)
siteid

PRODUCT  VALUE  DATE_TIME  DW_A_NUMBER  DW_B_NUMBER  CELL_ID
SITE_ID                                                                 
b47a998a3e      136    136        136          136          136      136
a7bd4c027b       85     85         85           85           85       85
7f76d66503       82     82         82           82           82       82
928a59b635       79     79         79           79           79       79
5c565fbfb1       76     76         76           76           76       76
...             ...    ...        ...          ...          ...      ...
cf2d02ac98        1      1          1            1            1        1
86a16450ae        1      1          1            1            1        1
d1251e1704        1      1          1            1            1        1
5083443d36        1      1          1            1            1        1
J2VnMbxAE+        1      1          1            1            1        1

[1281 rows x 6 columns]

In [74]:
#the top 10 most used sites in th three days
siteid.head(20)

PRODUCT  VALUE  DATE_TIME  DW_A_NUMBER  DW_B_NUMBER  CELL_ID
SITE_ID                                                                 
b47a998a3e      136    136        136          136          136      136
a7bd4c027b       85     85         85           85           85       85
7f76d66503       82     82         82           82           82       82
928a59b635       79     79         79           79           79       79
5c565fbfb1       76     76         76           76           76       76
91e86be9a6       71     71         71           71           71       71
c07a544c63       69     69         69           69           69       69
47657c9634       69     69         69           69           69       69
bc58a52936       66     66         66           66           66       66
ebfc58af70       62     62         62           62           62       62
d06f40f9d5       61     61         61           61           61       61
d1809b909f       61     61         61           61           61       61
dac8dcaeb9       60     60         60           60           60       60
487765ff90       59     59         59           59           59       59
8fd6f8d1c5       59     59         59           59           59       59
d96dcd5475       59     59         59           59           59       59
585a3bc66c       57     57         57           57           57       57
9bb4b27356       54     54         54           54           54       54
c1ccd3443b       52     52         52           52           52       52
0f34c36ac5       51     51         51           51           51       51

In [75]:
siteid.sum()

PRODUCT        12895
VALUE          12895
DATE_TIME      12895
DW_A_NUMBER    12895
DW_B_NUMBER    12895
CELL_ID        12895
dtype: int64

In [76]:
#the 10 least used sites in the three days
siteid.tail(10)

PRODUCT  VALUE  DATE_TIME  DW_A_NUMBER  DW_B_NUMBER  CELL_ID
SITE_ID                                                                 
c1d42f5f3f        1      1          1            1            1        1
c58809bff2        1      1          1            1            1        1
ce1d1c2d2e        1      1          1            1            1        1
6e9d0c4534        1      1          1            1            1        1
12f8d90487        1      1          1            1            1        1
cf2d02ac98        1      1          1            1            1        1
86a16450ae        1      1          1            1            1        1
d1251e1704        1      1          1            1            1        1
5083443d36        1      1          1            1            1        1
J2VnMbxAE+        1      1          1            1            1        1

In [77]:
#grouping by time
time = dftel.groupby('DATE_TIME')
time2 =time.count()
time2

PRODUCT  VALUE  DW_A_NUMBER  DW_B_NUMBER  CELL_ID  \
DATE_TIME                                                                
2012-05-06 23:04:37        2      2            2            2        2   
2012-05-06 23:05:25        2      2            2            2        2   
2012-05-06 23:05:41        2      2            2            2        2   
2012-05-06 23:05:42        2      2            2            2        2   
2012-05-06 23:06:29        2      2            2            2        2   
...                      ...    ...          ...          ...      ...   
2012-05-08 00:01:49       42     42           42           42       42   
2012-05-08 00:01:50       39     39           39           39       39   
2012-05-08 00:01:51       37     37           37           37       37   
2012-05-08 00:01:52       36     36           36           36       36   
2012-05-08 00:01:53       36     36           36           36       36   

                     SITE_ID  
DATE_TIME                     
2012-05-06 23:04:37        2  
2012-05-06 23:05:25        2  
2012-05-06 23:05:41        2  
2012-05-06 23:05:42        2  
2012-05-06 23:06:29        2  
...                      ...  
2012-05-08 00:01:49       42  
2012-05-08 00:01:50       39  
2012-05-08 00:01:51       37  
2012-05-08 00:01:52       36  
2012-05-08 00:01:53       36  

[1496 rows x 6 columns]

In [78]:
#preview of the top 200 time frames
timesort = time2.sort_values('PRODUCT',ascending = False)
timesort.head(200)

PRODUCT  VALUE  DW_A_NUMBER  DW_B_NUMBER  CELL_ID  \
DATE_TIME                                                                
2012-05-07 00:00:05       92     92           92           92       92   
2012-05-07 00:01:41       92     92           92           92       92   
2012-05-07 00:00:06       88     88           88           88       88   
2012-05-07 00:01:32       88     88           88           88       88   
2012-05-07 00:00:20       84     84           84           84       84   
...                      ...    ...          ...          ...      ...   
2012-05-08 00:00:40       28     28           28           28       28   
2012-05-08 00:01:06       28     28           28           28       28   
2012-05-08 00:01:21       28     28           28           28       28   
2012-05-08 00:00:20       27     27           27           27       27   
2012-05-08 00:00:38       27     27           27           27       27   

                     SITE_ID  
DATE_TIME                     
2012-05-07 00:00:05       92  
2012-05-07 00:01:41       92  
2012-05-07 00:00:06       88  
2012-05-07 00:01:32       88  
2012-05-07 00:00:20       84  
...                      ...  
2012-05-08 00:00:40       28  
2012-05-08 00:01:06       28  
2012-05-08 00:01:21       28  
2012-05-08 00:00:20       27  
2012-05-08 00:00:38       27  

[200 rows x 6 columns]

In [79]:
#grouping by date
Date =dftel.resample('D', on = 'DATE_TIME')


In [80]:
Datecount = Date.count()
Datecount

PRODUCT  VALUE  DATE_TIME  DW_A_NUMBER  DW_B_NUMBER  CELL_ID  \
DATE_TIME                                                                  
2012-05-06     1324   1324       1324         1324         1324     1324   
2012-05-07     8012   8012       8012         8012         8012     8012   
2012-05-08     3559   3559       3559         3559         3559     3559   

            SITE_ID  
DATE_TIME            
2012-05-06     1324  
2012-05-07     8012  
2012-05-08     3559

In [81]:
#grouping by time per hour
hour =dftel.resample('H', on = 'DATE_TIME')

In [82]:
#Number of entries recorded per hour
hourcount = hour.count()
hourcount

PRODUCT  VALUE  DATE_TIME  DW_A_NUMBER  DW_B_NUMBER  \
DATE_TIME                                                                  
2012-05-06 23:00:00     1324   1324       1324         1324         1324   
2012-05-07 00:00:00     7246   7246       7246         7246         7246   
2012-05-07 01:00:00        0      0          0            0            0   
2012-05-07 02:00:00        0      0          0            0            0   
2012-05-07 03:00:00        0      0          0            0            0   
2012-05-07 04:00:00        0      0          0            0            0   
2012-05-07 05:00:00        0      0          0            0            0   
2012-05-07 06:00:00        0      0          0            0            0   
2012-05-07 07:00:00        0      0          0            0            0   
2012-05-07 08:00:00        0      0          0            0            0   
2012-05-07 09:00:00        0      0          0            0            0   
2012-05-07 10:00:00        0      0          0            0            0   
2012-05-07 11:00:00        0      0          0            0            0   
2012-05-07 12:00:00        0      0          0            0            0   
2012-05-07 13:00:00        0      0          0            0            0   
2012-05-07 14:00:00        0      0          0            0            0   
2012-05-07 15:00:00        0      0          0            0            0   
2012-05-07 16:00:00        0      0          0            0            0   
2012-05-07 17:00:00        0      0          0            0            0   
2012-05-07 18:00:00        0      0          0            0            0   
2012-05-07 19:00:00        0      0          0            0            0   
2012-05-07 20:00:00        0      0          0            0            0   
2012-05-07 21:00:00        0      0          0            0            0   
2012-05-07 22:00:00        0      0          0            0            0   
2012-05-07 23:00:00      766    766        766          766          766   
2012-05-08 00:00:00     3559   3559       3559         3559         3559   

                     CELL_ID  SITE_ID  
DATE_TIME                              
2012-05-06 23:00:00     1324     1324  
2012-05-07 00:00:00     7246     7246  
2012-05-07 01:00:00        0        0  
2012-05-07 02:00:00        0        0  
2012-05-07 03:00:00        0        0  
2012-05-07 04:00:00        0        0  
2012-05-07 05:00:00        0        0  
2012-05-07 06:00:00        0        0  
2012-05-07 07:00:00        0        0  
2012-05-07 08:00:00        0        0  
2012-05-07 09:00:00        0        0  
2012-05-07 10:00:00        0        0  
2012-05-07 11:00:00        0        0  
2012-05-07 12:00:00        0        0  
2012-05-07 13:00:00        0        0  
2012-05-07 14:00:00        0        0  
2012-05-07 15:00:00        0        0  
2012-05-07 16:00:00        0        0  
2012-05-07 17:00:00        0        0  
2012-05-07 18:00:00        0        0  
2012-05-07 19:00:00        0        0  
2012-05-07 20:00:00        0        0  
2012-05-07 21:00:00        0        0  
2012-05-07 22:00:00        0        0  
2012-05-07 23:00:00      766      766  
2012-05-08 00:00:00     3559     3559

EVALUATION

This analysis yeilded the following information that should be considered in the infrastructure upgrading strategy.
1. The most popular product used in the 3 days was voice followed by sms and then data.
   The total value(from billing information) gotten from each product was:
        Voice - 321,224
        sms - 39,659
        data - 1,999 
2. In terms of localisation about 41.72% of the data recorded was from Abidjan and the Interior recorded 
   about 58.27% of the data.
3. All the data was recorded within 3 days.
4. Of the 3days the 7th of May recorded the most entries which were 8,012 followed by 8th of May with 3,559
   and 6th May with 1,999
5. Most of the data recorded was between the hours of 2300HRS and midnight.